oop: adds intro-to-oop, intro-to-oop-contd, python-oop, oop-relationships. (#42)
Ada-Developers-Academy/core-unit-1: python-oop/static-methods-and-class-methods.md

I can access class variables using the cls parameter: This is an example class variable!
```

Of course, class variables can be accessed without the `cls` parameter, as long as there's access to the class itself. The following version of `ExampleClass` produces the same output as the previous version:

```python
class ExampleClass:
    example_class_var = "This is an example class variable!"

    @classmethod
    def example_method(cls):
        print("I'm inside the class method, example_method!")
        print("I can access class variables using the cls parameter:", ExampleClass.example_class_var)
```

The difference is that `cls` will _always_ represent the current class itself; specifying `ExampleClass` could have consequences if class definitions change.

## Class Methods: Application

Class methods are special because they have access to class variables, and class state.

Class methods are great for:

- modifying class state that applies across all instances of the class
- doing operations that aren't associated with an instance
- doing operations that rely on the class itself, such as making instances of that class

## Check for Understanding

> <!-- Use the static method decorator to create a static method that... -->
> <!-- Use the class method decorator to create a static method that... -->

Reviewer: ```suggestion
<!-- prettier-ignore-start -->
### !challenge

* type: checkbox
* id: 6d4dab23-bc9d-45fa-a4d8-07689ecf84c7
* title: Static Methods and Class Methods I 

##### !question
Select the options that complete this sentence:

> A static method ..
##### !end-question

##### !options
* is bound to a class
* does not have access to instance variables (attributes) or class variables. 
* can access or modify class state 
* requires `cls` as first argument
##### !end-options

##### !answer
* is bound to a class
* does not have access to instance variables (attributes) or class variables.
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge

* type: checkbox
* id: 6d4dab23-bc9d-45fa-a4d8-07689ecf84c7
* title: Static Methods and Class Methods II

##### !question
Select the options that complete this sentence:

> A class method ..
##### !end-question

##### !options
* is bound to a class
* has access to instance variables (attributes) or class variables. 
* can access or modify class state 
* requires `cls` as first argument
##### !end-options

##### !answer
* is bound to a class
* has access to instance variables (attributes) or class variables. 
* can access or modify class state 
* requires `cls` as first argument
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: multiple-choice
* id: 3c69fae9-f63c-42f4-915a-9fca215bc374
* title: Static Methods and Class Methods III

##### !question
Select the decorator that would call the `subtractNumbers` method and print `5` to the console.

``` Python
class Calculator:
    def subtractNumbers(x, y):
        return x - y
print(Calculator.subtractNumbers(10,5))
```
##### !end-question

##### !options
* @staticmethod
* @classmethod
##### !end-options

##### !answer
* @staticmethod
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: 1d5979f7-0617-4a16-88f7-2da87c6c5a75
* title: Static Methods and Class Methods IV

##### !question
Select the option that will print the following output:

```
Your total will be $20.
Ask about our membership program to receive a 10% discount today!
```
##### !end-question

##### !options
* 
``` Python
class CalculateTotal:
    summer_discount = 10
    winter_discount = 15
    @classmethod
    def promote_membership(cls):
        print(f"Ask about our membership program to receive a {cls.summer_discount}'%' discount today!")

def display_total(amount):
    print(f"You total will be ${amount}")
    print(CalculateTotal.promote_membership())

display_total(20)
```
* 
``` Python
class CalculateTotal:
    summer_discount = 10
    winter_discount = 15
    
    @classmethod
    def promote_membership(cls):
        return "Ask about our membership program to receive a " + str(CalculateTotal.summer_discount) + "% discount today!"

def display_total(amount):
    print(f"Your total will be ${amount}")
    print(CalculateTotal.promote_membership())

display_total(20)
```

##### !end-options

##### !answer
* 
``` Python
class CalculateTotal:
    summer_discount = 10
    winter_discount = 15
    @classmethod
    def promote_membership(cls):
        print(f"Ask about our membership program to receive a {cls.summer_discount}'%' discount today!")

def display_total(amount):
    print(f"You total will be ${amount}")
    print(CalculateTotal.promote_membership())

display_total(20)
```
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->
```

Reviewer 2: Problem 2 (Static Methods and Class Methods II):
- includes "has access to instance variables (attributes) or class variables." but it does _not_ have access to instance variables.  This should just be "has access to class variables, but not instance variables."

Problem 4 (Static Methods and Class Methods IV):
- has a typo in the answer output "You"
- has weird stuff going on with an extra print in the classmethod
- the "wrong" answer prints the same correct answer string

## Introduction

The most common kind of method defined in OOP is instance methods. These methods can only be called from an instance. Instance methods typically are used to describe behaviors that instances can do, and to update the state associated with one instance.

> There are other kinds of methods that we can define within a class. These methods will not be called from an instance, and won't be responsible for maintaining an instance.

Author: nice nice nice!

### Static Methods: Syntax Details

In Python, we actually _can_ call static methods from an object instance, not just a class. However, because we don't pass in `self`, we don't have a reference to `self`. Static methods do not depend on an instance.

Observe this code, which shows that we _can_ call static methods from an object instance, but we can't access `self`.

```python
class ExampleClass:

    def __init__(self, name):
        self.name = name

    @staticmethod
    def example_method():
        print("I'm inside the static method, example_method!")
        print(f"I'm accessing an instance variable, {self.name}")
```

```python
example_instance = ExampleClass("Hello, World!")

example_instance.example_method()
```

This code produces the console output:

```
I'm inside the static method, example_method!
Traceback (most recent call last):
  File "main.py", line 13, in <module>
    example_instance.example_method()
  File "main.py", line 9, in example_method
    print(f"I'm accessing an instance variable, {self.name}")
NameError: name 'self' is not defined
```

We should observe from the output that:

> - The code _did_ successfully begin the static method, and it _did_ print `I'm inside the static method, example_method!`
> - The code threw a runtime error `NameError`, because we are unable to access `self`

Reviewer 2: ```suggestion
- The code _did_ successfully begin the static method, and it _did_ print `I'm inside the static method, instanceish_method!`
- The code threw a runtime error `NameError`, because we are unable to access `self` since there is no self parameter.
```

> 

Reviewer 2: Another insertion with code

============================

Similarly, if we try to call `selfish_method` either on the class or the instance, like one of the following:

```python
ExampleClass.selfish_method()
example_instance.selfish_method()
```

We'll get an error like this:

```text
Traceback (most recent call last):
  File "main.py", line 23, in <module>
    example_instance.selfish_method()
TypeError: selfish_method() missing 1 required positional argument: 'self'
```

Our method didn't even get called! It expected one parameter, but none was passed. Interestingly, this implies that if we tried one of _these_ calls:

```python
ExampleClass.selfish_method(example_instance)
example_instance.selfish_method(example_instance)
```

It should work. It's not really useful for much of anything, but sure enough, if we try it then we get something like:

```text
I'm inside the static method, selfish_method!
Who am I? <__main__.ExampleClass object at 0x10e982310>
```

In this case, there's nothing special about `self`. It was a parameter that we passed in ourselves. So static methods really are just like regular functions. They just happen to be stored in a class, and accessible either through the class, or any instance of that class.

## Static Methods: Application

Static methods are often utility functions. "Utility functions" is an informal term to describe a function responsible for doing common, routine tasks.

Static methods are great for:

- returning fixed values not associated with an instance
- doing operations that aren't associated with an instance

## Class Methods are Methods with Class State

Class methods are methods that are only called from a class, not an instance.

Class methods have access to class variables. However, they do _not_ have access to any instance variables.

## Class Methods: Syntax

To define a class method, we:

- use the `@classmethod` decorator that the Python library provides

> - do not pass in `self` as a parameter
> - pass in `cls` as a parameter

Reviewer 2: we never pass self, we just add a parameter to receive it
```suggestion
- do _not_ include a parameter to receive `self`
- instead, include a parameter to receive the class, which by convention we call `cls`
```

> Instead of passing in `self`, the variable that represents an instance itself, in class methods, we pass in `cls`, a variable that represents the class itself.
> 
> `cls` is the conventional name for this parameter.
> 

Reviewer 2: I think stating this one in the bullets is fine since the example comes right away.
```suggestion
```

```python
class ExampleClass:

    @classmethod
    def example_method(cls):
        print("I'm inside the class method, example_method!")
        print("In a class method, cls will be the class itself", cls)
```

Once a static method is defined, we can invoke it using dot notation, typically from the class itself.

```python
ExampleClass.example_method()
```

This code produces the console output:

```
I'm inside the class method, example_method!
In a class method, cls will be the class itself <class '__main__.ExampleClass'>
```

> Any code that can reference the class can call a class method. Similar to static methods, we can call a class method from other static, class, or instance methods, or even inside another class.

Reviewer 2: similar change as before
```suggestion
Any code that can reference the class can call a class method. Similar to static methods, we can call a class method from other static, class, or instance methods, or even from inside another class.
```

### Class Methods: Syntax Details

We actually _can_ call class methods from an object instance, too. However, the `@classmethod` decorator will specify that the first parameter, `cls`, will _always_ have the value of the class itself.

> Even if we call a class method from an instance, we will not have access to the instance it`self`.

Reviewer 2: I see what you did there.

### Using `cls` to Access Class Variables

As mentioned above, class methods can access any class variables that are defined, using the `cls` variable:

```python
class ExampleClass:
    example_class_var = "This is an example class variable!"

    @classmethod
    def example_method(cls):
        print("I'm inside the class method, example_method!")
        print("I can access class variables using the cls parameter:", cls.example_class_var)
```

```
ExampleClass.example_method()
```

This code produces the console output:

```
I'm inside the class method, example_method!
I can access class variables using the cls parameter: This is an example class variable!
```

Of course, class variables can be accessed without the `cls` parameter, as long as there's access to the class itself. The following version of `ExampleClass` produces the same output as the previous version:

```python
class ExampleClass:
    example_class_var = "This is an example class variable!"

    @classmethod
    def example_method(cls):
        print("I'm inside the class method, example_method!")
        print("I can access class variables using the cls parameter:", ExampleClass.example_class_var)
```

> The difference is that `cls` will _always_ represent the current class itself; specifying `ExampleClass` could have consequences if class definitions change.

Reviewer 2: add stipulation about inheritance
```suggestion
The difference is that `cls` will _always_ represent the current class itself; specifying `ExampleClass` could have consequences if class definitions change, or when inheritance is involved.
```
